### ISSUE TITLE
Store action-specific state in `ActionState`

### ISSUE THREAD

**Issue by alice-i-cecile** [Leafwing-Studios/leafwing-input-manager#60]
Many actions are only meaningful in association with non-enumerable data.

A very useful pattern here is to store information in the enum variant, and then update that appropriately.

We want to be able to express:
```rust
#[derive(Actionlike, Clone)]
enum ArpgAction {
    Movement(Direction),
    Ability1,
    Ability2,
    Ability3,
    Ability4,
    Ultimate,
}

input_map.insert(ArpgAction::Movement(Direction::NORTH), KeyCode::Up);

let ArpgAction::Movement(direction) = action_state.get_value::<ArpgAction::Movement>();
```

Unfortunately, a reasonable API for the last step seems to be completely blocked on https://github.com/rust-lang/rfcs/pull/2593, which, unsurprisingly, is postponed indefinitely 🙃

**Comment by alice-i-cecile** [comment]
Blocks #50.

**Comment by alice-i-cecile** [comment]
We can eliminate the `Hash` bound by storing a `HashMap<usize, InputAction>` in `InputMap`, and doing likewise for `ActionMap`. This usize is given by `Actionlike::index()`, which has the desired behavior!

This may make #68 harder, but this feature is significantly more important, and serves related goals.

**Comment by alice-i-cecile** [comment]
We may be able to workaround this by generating an internal type to match those of each variant, and then requesting those in `ActionState::get_value` :/ It's *extremely* cursed though due to the high levels of magic, and will do wonky things with the imports.

**Comment by alice-i-cecile** [comment]
This is fundamentally wrong-headed, and not our responsibility.

**Comment by alice-i-cecile** [comment]
#151 has some related mechanisms, we should figure out how to steal from the approach used to store gamepad axis values.

**Comment by alice-i-cecile** [comment]
Closing this out because we can now store action state values.

**Comment by bardt** [comment]
@alice-i-cecile could you (or someone else) point to an example of how to store action state value? Or at least to the feature you mean.

I want to do something similar to the example in the issue description: pass additional data with an Actionlike enum variant.

**Comment by alice-i-cecile** [comment]
The feature that I've been using is the `DualAxis` / `SingleAxis` features, where we can store 2/1 f32s. Is that enough for your use case?

I'm not fully satisfied with the design here either/

**Comment by bardt** [comment]
I don't think that's enough for my use case, but to be fair, I'm trying to do a pretty weird thing. I move the mouse cursor on a tile grid and wanted to model this as an action, something like: 

```
enum Action {
	MoveTileCursor(GridCoords),
	Move, // to the tile cursor
	Shoot, // to the place under the tile cursor
}
```

Shoot and Move would map to mouse keys, but for `MoveTileCursor` I would need a custom system to listen to mouse events and translate them to the current GridCoords, and call `action_state.press(MoveTileCursor(grid_coords))`.

Why would I want to do this? To simplify AI. The AI opponent would do something like:

```
action_state.press(MoveTileCursor(human_player_grid_coords));
action_state.press(Shoot);
```

Am I completely off with my thinking?

**Comment by alice-i-cecile** [comment]
Nope, this is a very natural pattern, and something I'd considered before. Reopening this so we remember to chew on additional designs.
